### PR TITLE
Enable HomeFinderFeature as part of Graal SDK

### PR DESCRIPTION
**Pull request by jaikiran**
Fixes https://github.com/oracle/graal/issues/1889

The commit here moves the `HomeFinderFeature` from `sdk` project to the `substratevm` project and marks that class as a `AutomaticFeature`. The move from `sdk` to `substratevm` was necessary so that it could reference the `@AutomaticFeature` annotation.

Just like I noted in https://github.com/oracle/graal/issues/1855, I'm actually a bit surprised that the existing test cases did not catch this issue. In fact, there's a `org.graalvm.home.test.VersionTest` in `sdk` project which fails with the exact same exception as reported in https://github.com/oracle/graal/issues/1889 when run as `mx native-unittest org.graalvm.home.test.VersionTest` (I didn't even know there's a `native-unittest` goal/target in the mx build but just happened to find it in some other unrelated search):

```
Building junit image for matching: org.graalvm.home.test.VersionTest
Build on Server(pid: 33168, port: 64851)*
[com.oracle.svm.junit.svmjunitrunner:33168]    classlist:   2,760.39 ms
[com.oracle.svm.junit.svmjunitrunner:33168]        (cap):   2,289.95 ms
[com.oracle.svm.junit.svmjunitrunner:33168]        setup:   4,360.38 ms
[com.oracle.svm.junit.svmjunitrunner:33168]     analysis:   7,306.97 ms
Fatal error: com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing org.graalvm.home.HomeFinder.getInstance() 
Parsing context:
	parsing org.graalvm.home.Version.getCurrent(Version.java:293)
	parsing org.graalvm.home.test.VersionTest.testCurrent(VersionTest.java:153)
	parsing com.oracle.svm.reflect.VersionTest_testCurrent_88eeae51be26187b84de158b10b96158c2b791cc_3.invoke(Unknown Source)
	parsing java.lang.reflect.Method.invoke(Method.java:498)
	parsing org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	parsing org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	parsing org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:263)
	parsing org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	parsing org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	parsing org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	parsing com.oracle.mxtool.junit.MxJUnitWrapper$1.schedule(MxJUnitWrapper.java:280)
	parsing org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	parsing org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	parsing org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	parsing org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	parsing org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	parsing org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	parsing com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:310)
	parsing com.oracle.svm.junit.SVMJUnitRunner.run(SVMJUnitRunner.java:164)
	parsing com.oracle.svm.junit.SVMJUnitRunner.main(SVMJUnitRunner.java:184)
	parsing com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:151)
	parsing com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:186)
	parsing com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

	at com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:138)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:327)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:300)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:107)
	at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:346)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:511)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:171)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: com.oracle.svm.core.util.UserError$UserException: ImageSingletons do not contain key org.graalvm.home.HomeFinder
	at com.oracle.svm.core.util.UserError.abort(UserError.java:65)
	at com.oracle.svm.hosted.ImageSingletonsSupportImpl$HostedManagement.doLookup(ImageSingletonsSupportImpl.java:119)
	at com.oracle.svm.hosted.ImageSingletonsSupportImpl.lookup(ImageSingletonsSupportImpl.java:44)
	at org.graalvm.nativeimage.ImageSingletons.lookup(ImageSingletons.java:86)
	at com.oracle.svm.hosted.snippets.SubstrateGraphBuilderPlugins$42.apply(SubstrateGraphBuilderPlugins.java:894)
	at org.graalvm.compiler.nodes.graphbuilderconf.InvocationPlugin.execute(InvocationPlugin.java:213)
	at org.graalvm.compiler.java.BytecodeParser.tryInvocationPlugin(BytecodeParser.java:2097)
	at com.oracle.svm.hosted.phases.AnalysisGraphBuilderPhase$AnalysisBytecodeParser.tryInvocationPlugin(AnalysisGraphBuilderPhase.java:67)
	at org.graalvm.compiler.java.BytecodeParser.appendInvoke(BytecodeParser.java:1806)
	at org.graalvm.compiler.java.BytecodeParser.genInvokeStatic(BytecodeParser.java:1571)
	at org.graalvm.compiler.java.BytecodeParser.genInvokeStatic(BytecodeParser.java:1551)
	at org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5177)
	at org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3286)
	at org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3093)
	at org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:977)
	at org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:871)
	at org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:197)
	at org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:221)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:340)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:310)
	... 10 more
Error: Image build request failed with exit status 1
```

This `org.graalvm.home.test.VersionTest` now passes with this change:
```
Building junit image for matching: org.graalvm.home.test.VersionTest
Build on Server(pid: 36724, port: 65317)*
[com.oracle.svm.junit.svmjunitrunner:36724]    classlist:   2,512.36 ms
[com.oracle.svm.junit.svmjunitrunner:36724]        (cap):   2,105.49 ms
[com.oracle.svm.junit.svmjunitrunner:36724]        setup:   4,084.28 ms
[com.oracle.svm.junit.svmjunitrunner:36724]   (typeflow):   5,375.52 ms
[com.oracle.svm.junit.svmjunitrunner:36724]    (objects):   3,625.39 ms
[com.oracle.svm.junit.svmjunitrunner:36724]   (features):     417.55 ms
[com.oracle.svm.junit.svmjunitrunner:36724]     analysis:   9,637.63 ms
[com.oracle.svm.junit.svmjunitrunner:36724]     (clinit):     210.11 ms
[com.oracle.svm.junit.svmjunitrunner:36724]     universe:     696.39 ms
[com.oracle.svm.junit.svmjunitrunner:36724]      (parse):   1,031.08 ms
[com.oracle.svm.junit.svmjunitrunner:36724]     (inline):   1,612.38 ms
[com.oracle.svm.junit.svmjunitrunner:36724]    (compile):  12,069.46 ms
[com.oracle.svm.junit.svmjunitrunner:36724]      compile:  19,045.38 ms
[com.oracle.svm.junit.svmjunitrunner:36724]        image:   1,237.61 ms
[com.oracle.svm.junit.svmjunitrunner:36724]        write:     384.03 ms
[com.oracle.svm.junit.svmjunitrunner:36724]      [total]:  38,186.61 ms
Running: graal/substratevm/svmbuild/darwin-amd64/junit/tmpV2ReJI/com.oracle.svm.junit.svmjunitrunner --verbose
SVMJUnitCore
JUnit version 4.12
org.graalvm.home.test.VersionTest started (1 of 1)
  testParsing: Passed
  testDevSuffix: Passed
  testCompare: Passed
  testCurrent: Passed
  testSnapshot: Passed
  testCreate: Passed
  testEquals: Passed
org.graalvm.home.test.VersionTest finished
Time: 0.001

OK (7 tests)

Test run PASSED. Exiting with status 0.
```

### CODE REVIEW

File: sdk/src/org.graalvm.home/src/org/graalvm/home/HomeFinder.java [hunk 1]
```diff
@@ -108,10 +106,3 @@ public static HomeFinder getInstance() {
         }
     }
 }
-
-class HomeFinderFeature implements Feature {
```
Comment (eregon) [inline sdk/src/org.graalvm.home/src/org/graalvm/home/HomeFinder.java]: Maybe we should keep the class around, even if just empty, to not break potential external usages of `--features=org.graalvm.home.HomeFinderFeature`? I'm not sure.

**Comment by cstancu** [comment]
@jaikiran `mx native-unittest` by default only picks up tests in ['com.oracle.svm.test'](https://github.com/oracle/graal/blob/master/substratevm/mx.substratevm/mx_substratevm.py#L620), that's why it doesn't run `org.graalvm.home.test.VersionTest`.

File: sdk/src/org.graalvm.home/src/org/graalvm/home/HomeFinder.java (hunk 1 above)
Comment (cstancu) [reply sdk/src/org.graalvm.home/src/org/graalvm/home/HomeFinder.java]: Keeping around an empty feature class would be missleading.

**Comment by cstancu** [comment]
As for the package I think this should go into `com.oracle.svm.core.graal`.

**Comment by jaikiran** [comment]
> As for the package I think this should go into com.oracle.svm.core.graal.

Done - I've updated this PR to include this change.

**Comment by cstancu** [comment]
@jaikiran sorry for the delay. After further discussion with @olpaw we reached the conclusion that the best approach is to add a `native-image.properties` file in `graal-sdk.jar` containing `Args = --features=org.graalvm.home.HomeFinderFeature`. This way the `HomeFinderFeature` stays close to `HomeFinder`, the class whose behavior it alters. Moreover this approach allows the user to still opt out of using the automatic feature (which is not really possible with `@AutomaticFeature`) after https://github.com/oracle/graal/issues/1884 is implemented. If you don't have the time to make the change I'll do it myself to get this ready for 19.3.1.

**Comment by jaikiran** [comment]
Hello @cstancu, thank you for those details. I can work on this one and send an updated PR.

**Comment by jaikiran** [comment]
Hello @cstancu, I've updated this PR to incoporate your recent review suggestions and have introduced the `graal/sdk/src/org.graalvm.home/src/META-INF/native-image/org.graalvm.home/native-image.properties` to enable this `HomeFinderFeatue`.

> After further discussion with @olpaw we reached the conclusion that the best approach is to add a native-image.properties file in graal-sdk.jar containing Args = --features=org.graalvm.home.HomeFinderFeature

I looked at some of the existing `native-image.properties` files in the graal codebase and decided to use the `-H:Features=org.graalvm.home.HomeFinderFeature` construct to be consistent with the rest of the files which pass around these params using the `-H` construct. I hope that's OK.

I've run `mx native-unittest org.graalvm.home.test.VersionTest` with and without this change and I can confirm that the test now passes with this change (it fails without this change).

**Pull request merged**